### ISSUE TITLE
Excel: Current cell incorrectly announced when multiple cells selected

### ISSUE THREAD

**Issue by Qchristensen** [nvaccess/nvda#6959]
In Excel, when multiple cells are selected, one cell still has focus.  Usually, this is the first cell, and this is what NVDA announces.

It is possible to change the cell which has focus within a selection by pressing TAB or ENTER (focus will cycle through all selected cells).  If TAB is pressed, focus moves across the selected row, then down to the next selected row.  If ENTER is pressed, then focus will move down the selected column, then to the top of the next selected row.

If text is typed, the text in whichever cell currently has focus will be overwritten.  Whenever TAB or ENTER is pressed with multiple cells selected, NVDA reads the selection and then identifies the first cell.

For instance:
1. Select B3:D3.  NVDA announces "B3 through D3 selected, row 3, column 2".
2. Press TAB.  The three cells are still selected and focus moves to C3.  NVDA reads "B3 through D3 selected, row 3, column 2".

NVDA+TAB reads the selection again as in step 1 above.
NVDA+numpad Delete reads the position on the screen as a percentage.

Visually, the "name" box shows the currently focused cell.  Pressing alt+f3 moves the focus to this box and reads the currently focused cell.  Press ESCAPE to return to the sheet itself (without changing the selection).

Ideal outcomes:
- NVDA to read the currently focussed cell when TAB or ENTER is pressed, eg  "B3 through D3 selected, B3", then when TAB is pressed, read "B3 through D3 selected, C3".  I'm not sure if the original decision to read "Row 2, Column 3" rather than B3 was to avoid confusion?  In that case, Row 2, column 3, then press tab and have it read row 2, column 4 would work too.
- NVDA to read the currently focussed item when NVDA+TAB or NVDA+numpad Delete are pressed.

**Comment by jcsteh** [comment]
CC @MichaelDCurran. Can we even get the focused cell programmatically in
this case?

**Comment by Adriani90** [comment]
This issue is still reproducible in NVDA 2019.1 and Excel 2016 and 365.
I would prefer NVDA to report the cell with focus first and then the selected range. I.e.:
* B3, A3 through D3
After pressing tab
* C3, A3 through D3

**Comment by CyrilleB79** [comment]
I suspect #10540 being a duplicate of this one. See @k-kolev1985 's description in the ticket that may be interesting.

As written in #10540:
Note that in an Excel cell selection, you can move the focus the following way:
* Enter/Shift+Enter moves the focus in column and go to the next/previous column when the column end is reached.
* Tab/Shift+Tab moves the focus in line and go to the next/previous line when the line end is reached.

**Comment by Qchristensen** [comment]
Yes I just closed #10540 as a duplicate of this one, since much of the information from that issue has already been replicated here.  FWIW, still the same in Office 365 Version: 
16.0.12730.20256 and NVDA 2020.1

**Comment by CyrilleB79** [comment]
I confirm that the focused cell info is available. I have a try branch getting this info. I will provide a PR if people agree.
The point raised by @Adriani90 still need to be discussed: Do we need to announce only the info of the focused cell or also the selected range? If both information is announced, would it not be too verbose?

**Comment by Adriani90** [comment]
@CyrilleB79 I think it could be done like following:
* When sellecting cells, speak only the sellected cell range "a2 through b5"
* when pressing tab, report also the focused cell and afterwards the previously sellected cell range "c5 focused, a2 through b5 sellected".

**Comment by leonardder** [comment]
I think it makes sense to generalize this now this is also a discussion for LibreOffice.

Honestly, ideally the following should happen:

1. Selection is announced based on events not related to focus. In the current situation, we modify the focused object by changing the name and such, but this is pretty hacky and we should avoid this for selections.
2. NVDA should announce selection changes. When changing focus in a selected range, NVDA should only announce the focused cell. If the user wants an explicit selection change, the reportCurrentSelection command can do that.

**Comment by CyrilleB79** [comment]
> I think it makes sense to generalize this now this is also a discussion for LibreOffice.

I agree/

>     2. NVDA should announce selection changes. When changing focus in a selected range, NVDA should only announce the focused cell. If the user wants an explicit selection change, the reportCurrentSelection command can do that.

This is more logical and that was also my idea. And this would help people distinguish more clearly the difference between the focused cell and the selected cell(s) in Excel.

**Comment by michaelweghorn** [comment]
> Honestly, ideally the following should happen:
> 
>     1. Selection is announced based on events not related to focus. In the current situation, we modify the focused object by changing the name and such, but this is pretty hacky and we should avoid this for selections.
> 
>     2. NVDA should announce selection changes. When changing focus in a selected range, NVDA should only announce the focused cell. If the user wants an explicit selection change, the reportCurrentSelection command can do that.

That sounds good to me.
